include/wil/cppwinrt_authoring.h

```diff
@@ -330,5 +330,117 @@ struct single_threaded_notifying_property : single_threaded_rw_property<T>
+    using Xaml_PropertyMetadata = winrt::Microsoft::UI::Xaml::PropertyMetadata;
+#elif defined(WINRT_Windows_UI_Xaml_Data_H)
+    using Xaml_DependencyProperty = winrt::Windows::UI::Xaml::DependencyProperty;
+    using Xaml_PropertyChangedCallback = winrt::Windows::UI::Xaml::PropertyChangedCallback;
+    using Xaml_PropertyMetadata = winrt::Windows::UI::Xaml::PropertyMetadata;
+#endif
+
+    using Xaml_DependencyObject_GetValue = std::function<winrt::Windows::Foundation::IInspectable(Xaml_DependencyProperty const&)>;
+    using Xaml_DependencyObject_SetValue = std::function<void(Xaml_DependencyProperty const&, winrt::Windows::Foundation::IInspectable const&)>;
+
+    template <typename PropertyType, typename OwnerType>
+    inline Xaml_DependencyProperty register_dependency_property(
+        const std::wstring_view& propertyNameString)
+    {
+        return Xaml_DependencyProperty::Register(
+            propertyNameString,
+            winrt::template xaml_typename<PropertyType>(),
+            winrt::template xaml_typename<OwnerType>(),
+            nullptr);
+    }
+
+    template <typename PropertyType, typename OwnerType, typename DefaultValueType = PropertyType>
+    inline Xaml_DependencyProperty register_dependency_property(
+        const std::wstring_view& propertyNameString,
+        const DefaultValueType& defaultValue,
+        const Xaml_PropertyChangedCallback& propertyChangedCallback = nullptr)
+    {
+        // TODO: assert T and PropertyType are compatible
+        return Xaml_DependencyProperty::Register(
+            propertyNameString,
+            winrt::template xaml_typename<PropertyType>(),
+            winrt::template xaml_typename<OwnerType>(),
+            Xaml_PropertyMetadata{winrt::box_value(defaultValue), propertyChangedCallback});
+    }
+}
+
+//template <typename T>
+//struct single_threaded_dependency_property
+//{
+//    using Type = T;
+//
+//    single_threaded_dependency_property(
+//        wil::details::Xaml_DependencyProperty const& dp,
+//        wil::details::Xaml_DependencyObject_GetValue const& getValue,
+//        wil::details::Xaml_DependencyObject_SetValue const& setValue) :
+//        m_dp(dp), m_getValue(getValue), m_setValue(setValue)
+//    {
+//    }
+//
+//    template <typename Q>
+//    auto& operator()(Q&& q)
+//    {
+//        return winrt::unbox_value<Type>(m_getValue(m_dp));
+//    }
+//
+//    template <typename Q>
+//    auto& operator=(Q&& q)
+//    {
+//        m_setValue(m_dp, winrt::box_value(value));
+//    }
+//
+//private:
+//    wil::details::Xaml_DependencyProperty const& m_dp{ nullptr };
+//    wil::details::Xaml_DependencyObject_GetValue m_getValue{nullptr};
+//    wil::details::Xaml_DependencyObject_SetValue m_setValue{nullptr};
+//};
+
+#define WIL_DEFINE_DP(baseClass, type, name) \
+    static wil::details::Xaml_DependencyProperty name##Property() \
+    { \
+        static wil::details::Xaml_DependencyProperty s_##name##Property = \

```

Comment:
I chose to use function-local statics for a few reasons:

1. XAML doesn't seem to really care *when* these statics are initialized: in production, I've seen folks set the statics in the element's constructor, in an "Ensure" function that runs during app load, and even just in the root of the .cpp itself (which I know is dangerous).
2. I'm sure this differs from global statics in some respects (is this specific to the translation unit?)---but I'm not sure if that matters for this type of call. What is different because of this change?
3. Using function-local statics dramatically simplifies this code.

What should we do here?

Reply:
> which I know is dangerous

Do you mind explaining how this is dangerous? It's what I've mostly seen around.

> XAML doesn't seem to really care when these statics are initialized

Is this also true when e.g. the first instance of the class being created is through a `.xaml` file and that specific file assigns some DPs? I would've expected XAML to have difficulty there if lazily created since it is gonna try to assign properties that don't have registered metadata yet.

What I do here in my own macro is use C++17's `inline static`s, which end up emulating global static behavior but allows you to define it in the header (so I'm wondering if I should move away from this because your statement of them being dangerous).

Also, inline statics aren't specific to the TU.